optimizely/config_manager.py

```diff
@@ -0,0 +1,198 @@
+
+  def __init__(self,
+               datafile,
+               logger=None,
+               error_handler=None):
+    self.logger = logger or _logging.adapt_logger(logger or _logging.NoOpLogger())
+    self.error_handler = error_handler or noop_error_handler
+    self._datafile = datafile
+    self._config = project_config.ProjectConfig(self._datafile, self.logger, self.error_handler)
+
+  def get_config(self):
+    """ Returns instance of ProjectConfig.
+
+    Returns:
+      ProjectConfig.
+    """
+    return self._config
+
+
+class PollingConfigManager(BaseConfigManager):
+  """ Config manager that polls for the datafile and updated ProjectConfig based on an update interval. """
+
+  def __init__(self,
+               sdk_key=None,
+               update_interval=None,
+               url=None,
+               url_template=None,
+               logger=None,
+               error_handler=None):
+    """ Initialize config manager. One of sdk_key or url has to be set to be able to use.
+
+    Args:
+      sdk_key: Optional string uniquely identifying the datafile.
+      update_interval: Optional floating point number representing time interval in seconds
+                       at which to request datafile and set ProjectConfig.
+      url: Optional string representing URL from where to fetch the datafile. If set it supersedes the sdk_key.
+      url_template: Optional string template which in conjunction with sdk_key
+                    determines URL from where to fetch the datafile.
+      logger: Provides a logger instance.
+      error_handler: Provides a handle_error method to handle exceptions.
+    """
+    self.datafile_url = self.get_datafile_url(sdk_key, url, url_template)
+    self.update_interval = self.get_update_interval(update_interval)
+    self.last_modified = None
+    self._datafile = None
+    self._config = None
+    self._polling_thread = threading.Thread(target=self._run)
+    self.is_running = False
+    self.logger = logger or _logging.adapt_logger(logger or _logging.SimpleLogger(min_level=enums.LogLevels.DEBUG))
+    self.error_handler = error_handler or noop_error_handler
+
+  @staticmethod
+  def get_datafile_url(sdk_key, url, url_template):
+    """ Helper method to determine URL from where to fetch the datafile.
+
+    Args:
+      sdk_key: Key uniquely identifying the datafile.
+      url: String representing URL from which to fetch the datafile.
+      url_template: String representing template which is filled in with
+                    SDK key to determine URL from which to fetch the datafile.
+
+    Returns:
+      String representing URL to fetch datafile from.
+    """
+    # Ensure that either is provided by the user.
+    if sdk_key is None and url is None:
+      raise exceptions.InvalidInputException('Must provide at least one of sdk_key or url.')
+
+    url_template = url_template or enums.ConfigManager.DATAFILE_URL_TEMPLATE
```

Comment:
Does it make sense to just add `enums.ConfigManager.DATAFILE_URL_TEMPLATE` as a default value for the arg being passed in?

Reply:
Addressed it in caller itself.